### PR TITLE
Enable Android integration tests in remaining plugins

### PR DESCRIPTION
**Pull request by blasten**
Fixes https://github.com/flutter/flutter/issues/86749

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Looks like a number of these either aren't set up quite right or are failing tests; I think it would be easier to split them into one PR per plugin. (Looks like video_player worked, so that one we can land easily.)

**Comment by stuartmorgan** [comment]
I've been trying to do `shared_preferences` https://github.com/flutter/plugins/pull/4505, so if I can get that working we won't need a separate PR for it. I've failed quite a few times though (it's not clear to me how to run things locally the way FTL runs them, so iteration is very slow).

File: packages/camera/camera/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.9.4+5
```
Comment (stuartmorgan) [inline packages/camera/camera/CHANGELOG.md:1]: None of these need version bumps; you can just use `NEXT`. Test-only/example-only changes are exempt from version changes because they don't affect clients of the packages.

File: packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -5,4 +5,10 @@
       android:theme="@android:style/Theme.NoTitleBar.Fullscreen"
       android:exported="false"/>
   </application>
+  <queries>
+    <intent>
+      <action android:name="android.intent.action.SENDTO" />
```
Comment (blasten) [inline packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml:10]: This was causing an issue in Android 11

File: packages/camera/camera/CHANGELOG.md (hunk 1 above)
Comment (blasten) [reply packages/camera/camera/CHANGELOG.md]: Done

Comment (blasten) [reply packages/camera/camera/CHANGELOG.md]: After all, I fixed a bug here. I'm bumping the version, then

**Comment by blasten** [comment]
@stuartmorgan this is ready for another review

File: packages/in_app_purchase/in_app_purchase/example/integration_test/in_app_purchase_test.dart [hunk 3]
```diff
@@ -4,12 +4,14 @@
...
 import 'package:flutter_test/flutter_test.dart';
 import 'package:in_app_purchase/in_app_purchase.dart';
+import 'package:in_app_purchase_android/in_app_purchase_android.dart';
 import 'package:integration_test/integration_test.dart';
 
 void main() {
   IntegrationTestWidgetsFlutterBinding.ensureInitialized();
 
   testWidgets('Can create InAppPurchase instance', (WidgetTester tester) async {
+    InAppPurchaseAndroidPlatformAddition.enablePendingPurchases();
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase/example/integration_test/in_app_purchase_test.dart:14]: Why is this necessary for a placeholder test?

File: packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart [hunk 4]
```diff
@@ -17,10 +17,11 @@ void main() {
 
     // Generally all devices should have some default browser.
     expect(await canLaunch('http://flutter.dev'), true);
+    expect(await canLaunch('https:///www.google.com/404'), true);
 
     // SMS handling is available by default on most platforms.
     if (kIsWeb || !(Platform.isLinux || Platform.isWindows)) {
-      expect(await canLaunch('sms:5555555555'), true);
+      expect(await canLaunch('sms:+12345678900'), true);
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart:24]: `canLaunch` was giving phone-number-specific answers? That seems very strange.

File: packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml:10]: For which scheme? It sounds like we may need to update [the README](https://github.com/flutter/plugins/blob/master/packages/url_launcher/url_launcher/README.md#android).

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looks like `camera` is still broken. Per my earlier comment: is there a reason not to split that out so we can land the rest now while that's worked out?

File: packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (blasten) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml:10]: Android manifests are aggregated. There's no need for manual instructions.

File: packages/in_app_purchase/in_app_purchase/example/integration_test/in_app_purchase_test.dart (hunk 3 above)
Comment (blasten) [reply packages/in_app_purchase/in_app_purchase/example/integration_test/in_app_purchase_test.dart:14]: The constructor of `InAppPurchaseAndroidPlatformAddition` has an assert that requires this to be called. This is triggered when a new singleton is created via the instance getter.

File: packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart (hunk 4 above)
Comment (blasten) [reply packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart:25]: I added an extra test case. This isn't an answer. This is a test case to launch the Message app with this intent to send an SMS to the given phone number.

File: packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (blasten) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml:10]: SENDTO works for SMS and emails

Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml:10]: > Android manifests are aggregated. There's no need for manual instructions.

Oh, I thought this was in the example app. I don't think we should add this here, because we don't know what schemes someone is going to use. We should instead document this query in the README for people who need it, as we are currently doing (but without this scheme). And then add this to the example instead, so that the test passes.

File: packages/in_app_purchase/in_app_purchase/example/integration_test/in_app_purchase_test.dart (hunk 3 above)
Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase/example/integration_test/in_app_purchase_test.dart:14]: That doesn't seem ideal; can you file an issue about this and annotate the line with a TODO referencing it?

Also, don't we need to platform-gate this call?

File: packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart (hunk 4 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart:25]: > I added an extra test case. This isn't an answer.

I know it's a test case; the version I reviewed was changing the phone number of the existing case, not adding a new case. I assumed there was a reason for that, rather than it being a random change, and the only reason I could think of would be if `5555555555` was returning false but `+12345678900` were returning true.

> This is a test case to detect if the app can trigger an intent to send an SMS to the given phone number.

If they don't give different results, then I'm not sure what the purpose of the new test case is. Why do we need to test different number formats given that the plugin shouldn't care what the number format is (since it's the OS that parses that)?

Comment (blasten) [reply packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart]: I thought this was causing problems on Android 30, but it wasn't. The problem was a change in Android 30 about how `resolveActivity` works. 

I removed the test case.

File: packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (blasten) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml:10]: Done. I don't know why not adding them to the plugin AndroidManifest.xml automatically. I think the idea is to have 3p dependencies auto-configuring themselves when possible unless there's some major concern that prevents these from being the default.

I'm not sure if these concerns exist, or if we didn't know that the Android Gradle plugin merges manifests automatically.

Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml]: > I don't know why not adding them to the plugin AndroidManifest.xml automatically.

It is entirely possible to use `url_launcher` without any permissions (just use `launch` without `canLaunch`), and almost nobody would use the entire set of possible queries that `canLaunch` could require. We shouldn't force people using a plugin to have their app require permissions they aren't using. Especially since the guidance from the stores is always to use the minimal permission set.

> if we didn't know that the Android Gradle plugin merges manifests automatically

I didn't know that certainly, and it's good to know 🙂 . We should audit our READMEs for cases where the advice is "you must add X if you use this plugin" rather than "you might need X if you are using this plugin to do Y", and for the former, put it in the plugin manifest.

Comment (blasten) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml]: I haven’t used this plugin, but if canLaunch isn’t required, then the API seems very confusing. I assumed that canLaunch means that if it returns false, then never launch the intent, and handle it differently.

Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/android/src/main/AndroidManifest.xml]: Updating the docs is on my TODO list, because the example shown in the README (`await canLaunch(_url) ? await launch(_url) : throw 'Could not launch $_url';`) is pointless; it's just a slower and more cumbersome way of writing
```
if (!(await launch(_url))) throw 'Could not launch $_url';
```

> and handle it differently

This is where `canLaunch` is actually useful, yes. But with `https` URLs, for instance, I would be willing to bet that nobody has fallback code in their application that handles the case where someone's device *does not have a browser*, so calling `canLaunch` in that cirumstance is pointless.

File: packages/camera/camera/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,8 @@
+##  0.9.4+5
+
+* Fixes bug when calling method after the camera was closed resulted in a Java `IllegalStateException` exception.
```
Comment (stuartmorgan) [inline packages/camera/camera/CHANGELOG.md:3]: Nit: s/when calling method after/where calling a method after/

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 6]
```diff
@@ -281,8 +281,10 @@ public void onOpened(@NonNull CameraDevice device) {
           public void onClosed(@NonNull CameraDevice camera) {
             Log.i(TAG, "open | onClosed");
 
+            // Prevents calls to methods that would otherwise result in IllegalStateException exceptions.
+            cameraDevice = null;
+            closeCaptureSession();
             dartMessenger.sendCameraClosingEvent();
-            super.onClosed(camera);
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:285]: Why don't we need this?

File: packages/camera/camera/example/integration_test/camera_test.dart [hunk 7]
```diff
@@ -92,7 +92,9 @@ void main() {
         await controller.dispose();
       }
     }
-  }, skip: !Platform.isAndroid);
+  },
+      skip:
+          true); // TODO(egarciad): Fix https://github.com/flutter/flutter/issues/93686.
```
Comment (stuartmorgan) [inline packages/camera/camera/example/integration_test/camera_test.dart:97]: Optional nit for these: if you put the comment above the line rather than at the end of it, you can avoid the weird breaking between `skip` and `true`; that's how we've commented other similar cases.

File: packages/shared_preferences/shared_preferences/example/android/app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -1,5 +1,5 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="io.flutter.plugins.sharedpreferencesexample">
+    package="io.flutter.plugins.sharedpreferencesexample">
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences/example/android/app/src/main/AndroidManifest.xml:2]: This change (adding trailing whitespace) should be reverted.

File: packages/url_launcher/url_launcher/README.md [hunk 9]
```diff
@@ -72,7 +72,12 @@ for examples of other queries.
     <action android:name="android.intent.action.DIAL" />
     <data android:scheme="tel" />
   </intent>
-  <!-- If your app emails -->
+  <!-- If your sends SMS messages -->
+  <intent>
+    <action android:name="android.intent.action.SENDTO" />
+    <data android:scheme="smsto" />
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/README.md:78]: Is this a typo? I've only ever seen `sms` as the scheme.

File: packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart [hunk 10]
```diff
@@ -17,6 +17,7 @@ void main() {
 
     // Generally all devices should have some default browser.
     expect(await canLaunch('http://flutter.dev'), true);
+    expect(await canLaunch('https:///www.google.com/404'), true);
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart:20]: Why is there an extra `/` after the scheme here?

**Review by stuartmorgan — APPROVED** [review]
LGTM with some nits. Thanks for getting all of these working!

**Comment by stuartmorgan** [comment]
@blasten I wrote up https://github.com/flutter/flutter/wiki/Plugin-Tests#enabling-android-ui-tests based on everything I learned from my trial and error on `shared_preferences` and from your PR. Please edit it if you think anything is missing or misleading.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 6 above)
Comment (blasten) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:285]: The default implementation of `onClosed` does nothing.  For callbacks, I'd be very surprised if there was an API that has a super implementation that does something. It'd be incorrect to call it a callback.
 https://developer.android.com/reference/android/hardware/camera2/CameraDevice.StateCallback#onClosed(android.hardware.camera2.CameraDevice)

File: packages/camera/camera/CHANGELOG.md (hunk 5 above)
Comment (blasten) [reply packages/camera/camera/CHANGELOG.md:3]: done

File: packages/url_launcher/url_launcher/README.md (hunk 9 above)
Comment (blasten) [reply packages/url_launcher/url_launcher/README.md:78]: From the Android docs:

> 
> If you want to ensure that your intent is handled only by a text messaging app (and not other email or social apps), then use the ACTION_SENDTO action and include the "smsto:" data scheme. For example:
>    ` <data android:scheme="smsto" />`
>

File: packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart (hunk 10 above)
Comment (blasten) [reply packages/url_launcher/url_launcher/example/integration_test/url_launcher_test.dart:20]: Good catch. done

File: packages/camera/camera/example/integration_test/camera_test.dart (hunk 7 above)
Comment (blasten) [reply packages/camera/camera/example/integration_test/camera_test.dart]: done

File: packages/shared_preferences/shared_preferences/example/android/app/src/main/AndroidManifest.xml (hunk 8 above)
Comment (blasten) [reply packages/shared_preferences/shared_preferences/example/android/app/src/main/AndroidManifest.xml]: done

**Comment by blasten** [comment]
@stuartmorgan yup. That's pretty much it

File: packages/url_launcher/url_launcher/README.md (hunk 9 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/README.md:78]: TIL!
